File: 0017-letter-combinations-of-a-phone-number/0017-letter-combinations-of-a-phone-number.cpp
class Solution {
public:
void combi(int index,string digits,string temp,vector<string>& ans,unordered_map<char,string>map){
        if(index==digits.size()){
            ans.push_back(temp);
            return;
        }
        for(char c: map[digits[index]])
        {
            temp.push_back(c);
            combi(index+1,digits,temp,ans,map);
            temp.pop_back();
        }
    }
    vector<string> letterCombinations(string digits) {
        if(digits=="") return { };
        unordered_map<char,string>map={{'2',"abc"},{'3',"def"},{'4',"ghi"},{'5',"jkl"},{'6',"mno"},{'7',"pqrs"},{'8',"tuv"}
        ,{'9',"wxyz"}};
        vector<string>ans;
        combi(0,digits,"",ans,map);
        return ans;        
    }
};

File: 0017-letter-combinations-of-a-phone-number/0017-letter-combinations-of-a-phone-number.java
class Solution {
    public final String[] m = {
        "", "",
        "abc", "def", "ghi",
        "jkl", "mno", "pqrs",
        "tuv", "wxyz"
    };
    public void f(int index,String digits,StringBuilder sb,List<String> ans){
        if(index == digits.length()){
            ans.add(sb.toString());
            return;
        }

        String s = m[digits.charAt(index) - '0'];
        for (int i = 0; i < s.length(); i++) {

        sb.append(s.charAt(i));

        f(index + 1, digits, sb, ans);

        sb.deleteCharAt(sb.length() - 1); 
        }

    }
    public List<String> letterCombinations(String digits) {
        List<String> ans = new ArrayList<>();
        if(digits.length() == 0){
            return ans;
        }
        
        f(0,digits,new StringBuilder(),ans);

        return ans;        
    }
}